src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs

```diff
@@ -0,0 +1,419 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.Net.Quic.Implementations.MsQuic.Internal;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+using static System.Net.Quic.Implementations.MsQuic.Internal.MsQuicNativeMethods;
+
+namespace System.Net.Quic.Implementations.MsQuic
+{
+    internal sealed class MsQuicConnection : QuicConnectionProvider
+    {
+        // Functions to invoke in MsQuic
+        public MsQuicApi _api;
+        private MsQuicSession _session;
+
+        // Pointer to the underlying connection
+        private IntPtr _ptr;
+
+        // Handle to this object for native callbacks.
+        private GCHandle _handle;
+
+        // Delegate that wraps the static function that will be called when receiving an event.
+        private ConnectionCallbackDelegate _connectionDelegate;
```

Comment:
Should this be static?

Reply:
I was hitting issues when this was static before, which was odd. Effectively after calling into the function a few thousand times, it would eventually AV saying the function is invalid. 

I can leave it as a TODO to investigate if this function can be static?

Reply:
Weird. Yeah, we should at least have a TODO.